### ISSUE TITLE
Getting "Code block produced no output" when a unittest module test fails

### ISSUE THREAD

**Issue by kaushalmodi** [Lompik/ob-nim#4]
Hello,

Here is an Org file to reproduce the issue.

```org
#+title: No ob-nim output when unittest fails
#+date: <2018-05-25 Fri>

#+property: header-args :eval never-export
#+property: header-args:nim :exports both :results output replace verbatim
* Passing test
#+begin_src nim :tangle test_passing.nim
when isMainModule:
  import unittest
  suite "check number equality":

    test "1 == 1":
      check:
        1 == 1
#+end_src

#+RESULTS:
:
: [Suite] check number equality
:   [OK] 1 == 1

* Failing test
#+begin_src nim :tangle test_failing.nim
when isMainModule:
  import unittest
  suite "check number equality (failing)":

    test "1 == 0":
      check:
        1 == 0
#+end_src

#+RESULTS:
```

- In the passing test, the `unittest` generated outputs get inserted fine in the `#+RESULTS` block.
- But when the test fails, I just get "Code block produced no output." in the echo area. I neither get the results block updated, not get the "\*Org-Babel Error Output\*" buffer.

Now, with the point in the code block in the "Failing test" subtree, if you tangle it using `C-c C-v t`, it will generate `test_failing.nim` file.

When you run that using:

```
nim c -r test_failing.nim
```

you will get:

```
[Suite] check number equality (failing)
    test_failing.nim(7, 10): Check failed: 1 == 0
  [FAILED] 1 == 0
```

It would be great if you can help debug and fix this.

Thanks!

**Comment by kaushalmodi** [comment]
Some debug.. If I modify the `org-babel-nim-execute` function by adding the below 2 `message` forms:

```elisp
    (let ((results
	   (org-babel-eval
	    (concat tmp-bin-file cmdline)
	    "")))
      (message "command: %s" (concat tmp-bin-file cmdline))
      (message "results: %S" results)
```

I see the the printed results is *nil*, and the command is something like `/tmp/babel-34Cxet/nim_src_EkwHIi`, and when I execute that binary manually, I indeed get the error on stdout.. 

```
km²/tmp:babel-34Cxet> ./nim_src_EkwHIi
[Suite] check number equality (failing)
    nim_src_EkwHIi.nim(10, 10): Check failed: 1 == 0
  [FAILED] 1 == 0
```

Now I am wondering why `org-babel-eval` was unable to catch that..

**Comment by Lompik** [comment]
you're right, ```org-babel-eval``` does not catch stdout. I added ```(cmdline (concat cmdline " 1>&2"))``` to ```org-babel-nim-execute``` and I could see the output in the ```*Org-Babel Error Output*``` buffer. 

there are workarounds but shouldn't nim use ```stderr``` on errors ?

**Comment by kaushalmodi** [comment]
So the `unittest` code is returning the error code as 1 but sending the output on stdout instead of stderr?

Copying @zah, the author of `unittest` for more info.

@zah: Hello, sorry to drag you into this... I was actually documenting the use of `unittest` for my learning, and I use Org mode + this `ob-nim` package by @Lompik -- See https://scripter.co/notes/nim/#failure-output.

Turns out that it is able to "see" the output on stdout when a `unittest` test passes.. 

But when error happens (like you see in that test code in my first post in this thread), Org see the return code of 1, but apparently nothing on the stderr. Is that right?

**Comment by kaushalmodi** [comment]
@Lompik 

> I added (cmdline (concat cmdline " 1>&2")) to org-babel-nim-execute and I could see the output.

Thanks! I confirm that.

> but shouldn't nim use stderr on errors ?

All other Nim errors like the ones generated during compilation or via `assert` show up in the error buffer fine.

Only the `unittest` errors seem to disappear, and they appear neither in the error buffer nor in the results buffer.

**Comment by kaushalmodi** [comment]
@Lompik 

Actually you are right.. even running this simple snippet doesn't get the error caught in the Org babel error buffer or the results drawer:

```nim
stderr.writeLine "Error!"
```

.. and with the `(cmdline (concat cmdline " 1>&2"))` hack, the regular stdout results somehow stop showing up in the results drawer.

:(

**Comment by Lompik** [comment]
try with 

    stderr.writeLine "Error!"
    quit 1

should work out of the box.

**Comment by kaushalmodi** [comment]
@Lompik Thanks I confirm that.. I understood my mistake... I was sending a message to *stderr* but returning a *success* return code.

`unittest` is doing the opposite.. sending messages to *stdout* but returning an *error* return code.. and so in both cases, neither the results drawer caught the messages nor the Org Babel Error buffer.

Can you please leave this issue open till @zah responds?
